### ISSUE TITLE
Repeated crashes in Nightly build, 3.00+ (33553c5331) (macOS)

### ISSUE THREAD

**Issue by ViRo3** [transmission/transmission#2365]
[Transmission-2021-12-27-205007.ips.txt](https://github.com/transmission/transmission/files/7792815/Transmission-2021-12-27-205007.ips.txt)
[Transmission-2021-12-27-205415.ips.txt](https://github.com/transmission/transmission/files/7792816/Transmission-2021-12-27-205415.ips.txt)
[Transmission-2021-12-30-163221.ips.txt](https://github.com/transmission/transmission/files/7792817/Transmission-2021-12-30-163221.ips.txt)
[Transmission-2021-12-30-164120.ips.txt](https://github.com/transmission/transmission/files/7792818/Transmission-2021-12-30-164120.ips.txt)

Attached above are the logs for crashes.

**Comment by ckerr** [comment]
> tr_assert_report(char const*, int, char const*, ...)","symbolLocation":112,"imageIndex":7},{"imageOffset":823968,"symbol":"tr_bitfield::countFlags(unsigned long, unsigned long) const","symbolLocation":324,"imageIndex":7},{"imageOffset":622848,"symbol":"tr_torrentGotBlock(tr_torrent*, unsigned int)","symbolLocation":448,"imageIndex":7},{"imageOffset":737476,"symbol":"peerCallbackFunc(tr_peer*, tr_peer_event const*, void*)","symbolLocation":744,"imageIndex":7},{"imageOffset":765260,"symbol":"canRead(tr_peerIo*, void*, unsigned long*)","symbolLocation":1088,"imageIndex":7},{"imageOffset":721232,"symbol":"canReadWrapper(tr_peerIo*)","symbolLocation":200,"imageIndex":7},{"imageOffset":719132,"symbol":"utp_on_read(void*, unsig

> "tr_assert_report(char const*, int, char const*, ...)","symbolLocation":112,"imageIndex":7},{"imageOffset":823968,"symbol":"tr_bitfield::countFlags(unsigned long, unsigned long) const","symbolLocation":324,"imageIndex":7},{"imageOffset":622848,"symbol":"tr_torrentGotBlock(tr_torrent*, unsigned int)","symbolLocation":448,"imageIndex":7},{"imageOffset":737476,"symbol":"peerCallbackFunc(tr_peer*, tr_peer_event const*, void*)","symbolLocation":744,"imageIndex":7},{"imageOffset":765260,"symbol":"canRead(tr_peerIo*, void*, unsigned long*)","symbolLocation":1088,"imageIndex":7},{"imageOffset":721232,"symbol":"canReadWrapper(tr_peerIo*)","symbolLocation":200,"imageIndex":7},{"imageOffset":719132,"symbol":"utp_on_read(void*, unsigned char const*, unsigned long)","symbolLocation":184,"imageIndex":7},{"imageOffset":1111224,"symbol":"UTP_ProcessIncoming(UTPSocket*, unsigned char const*, unsigned long, bool)"

> ,"symbol":"tr_assert_report(char const*, int, char const*, ...)","symbolLocation":112,"imageIndex":7},{"imageOffset":823968,"symbol":"tr_bitfield::countFlags(unsigned long, unsigned long) const","symbolLocation":324,"imageIndex":7},{"imageOffset":622848,"symbol":"tr_torrentGotBlock(tr_torrent*, unsigned int)","symbolLocation":448,"imageIndex":7},{"imageOffset":737476,"symbol":"peerCallbackFunc(tr_peer*, tr_peer_event const*, void*)","symbolLocation":744,"imageIndex":7},{"imageOffset":765260,"symbol":"canRead(tr_peerIo*, void*, unsigned long*)","symbolLocation":1088,"imageIndex":7}

**Comment by ViRo3** [comment]
additional logs incl spindump during crash
[s2.txt](https://github.com/transmission/transmission/files/7825906/s2.txt)
[Sample1.txt](https://github.com/transmission/transmission/files/7825908/Sample1.txt)
[spin1.txt](https://github.com/transmission/transmission/files/7825909/spin1.txt)
[spin2.txt](https://github.com/transmission/transmission/files/7825910/spin2.txt)

**Comment by ckerr** [comment]
@ViRo3 sorry for the delay on this. Is this still an issue in newer builds?

**Comment by ViRo3** [comment]
still happening for nested folders
[Transmission-2022-01-18-052243.ips.txt](https://github.com/transmission/transmission/files/7885216/Transmission-2022-01-18-052243.ips.txt)

**Comment by ckerr** [comment]
Here's the codepath causing the crash in `Transmission-2022-01-18-052243.ips.txt`

```
      {
        "imageOffset": 496404,
        "symbol": "abort",
        "symbolLocation": 164,
        "imageIndex": 6
      },
      {
        "imageOffset": 881196,
        "symbol": "tr_assert_report(char const*, int, char const*, ...)",
        "symbolLocation": 112,
        "imageIndex": 7
      },
      {
        "imageOffset": 824892,
        "symbol": "tr_bitfield::countFlags(unsigned long, unsigned long) const",
        "symbolLocation": 324,
        "imageIndex": 7
      },
      {
        "imageOffset": 623300,
        "symbol": "tr_torrentGotBlock(tr_torrent*, unsigned int)",
        "symbolLocation": 444,
        "imageIndex": 7
      },
      {
        "imageOffset": 739224,
        "symbol": "peerCallbackFunc(tr_peer*, tr_peer_event const*, void*)",
        "symbolLocation": 740,
        "imageIndex": 7
      },
      {
        "imageOffset": 767068,
        "symbol": "canRead(tr_peerIo*, void*, unsigned long*)",
        "symbolLocation": 1088,
        "imageIndex": 7
      },
```

**Comment by ckerr** [comment]
There are three assertions in `tr_bitfield::countFlags()` that this might be:

```
    TR_ASSERT(begin < end);
    TR_ASSERT(!std::empty(flags_));

    // and then, later in the function
    TR_ASSERT(ret <= (begin - end));
```

Most likely candidate is the third one, both because it' codepaths are more complex, and from the `symbolLocation: 324` that indicates the assertion occurs further into `countFlags()` than the first two asserts appear to be. When the proposal that @fxcoudert has in #2422 lands, that will remove this ambiguity.

I've got no idea how to reproduce this issue.

**Comment by ckerr** [comment]
@ViRo3 is this still happening in the newer nightlies? If so, the new crash reports should have more useful information because of #2422.

**Comment by ViRo3** [comment]
No it hasn't occured once since then. Do you rec closing the issue?

**Comment by ckerr** [comment]
Yeah. If it happens again, please do file a new issue with a crash report with that #2422 goodness
